pkg/apis/meta/v1alpha1/net.go

```diff
@@ -9,49 +10,95 @@ import (
 
 // PortMapping defines a port mapping between the VM and the host
 type PortMapping struct {
-	HostPort uint64 `json:"hostPort"`
-	VMPort   uint64 `json:"vmPort"`
+	BindAddress net.IP   `json:"bindAddress,omitempty"`
+	HostPort    uint64   `json:"hostPort"`
+	VMPort      uint64   `json:"vmPort"`
+	Protocol    Protocol `json:"protocol,omitempty"`
 }
 
 var _ fmt.Stringer = PortMapping{}
 
 func (p PortMapping) String() string {
-	return fmt.Sprintf("0.0.0.0:%d->%d", p.HostPort, p.VMPort)
+	var sb strings.Builder
+
+	if p.BindAddress != nil {
+		sb.WriteString(p.BindAddress.String())
+	} else {
+		sb.WriteString("0.0.0.0")
+	}
+
+	sb.WriteString(fmt.Sprintf(":%d->%d", p.HostPort, p.VMPort))
+
+	if len(p.Protocol) > 0 {
+		sb.WriteString(fmt.Sprintf("/%s", p.Protocol))
+	}
+
+	return sb.String()
 }
 
 // PortMappings represents a list of port mappings
 type PortMappings []PortMapping
 
 var _ fmt.Stringer = PortMappings{}
 
+var errInvalidPortMappingFormat = fmt.Errorf("port mappings must be of form [<bind address>:]<host port>:<VM port>[/<protocol>]")
+
 func ParsePortMappings(input []string) (PortMappings, error) {
```

Comment:
I'm strongly of the opinion that we should not duplicate code here that we already have vendored.
Please use https://github.com/docker/go-connections/blob/master/nat/nat.go#L126 for this task; and turn `map[Port][]PortBinding` from there into `meta.PortMappings`.

Reply:
Done in a182160.